examples/hyperparam/search_random.py

```diff
@@ -89,47 +84,37 @@ def get_metric(metric_name):
-        best_test_loss = math.inf
-        best_run = None
         with ThreadPoolExecutor(max_workers=max_p) as executor:
-            result = executor.map(new_eval(epochs,
+            _ = executor.map(new_eval(epochs,
                                            experiment_id,
                                            null_train_loss,
                                            null_val_loss,
                                            null_test_loss),
                                   runs)
-        tmp = tempfile.mkdtemp()
-        results_file_path = os.path.join(str(tmp), "results.txt")
-        with open(results_file_path, "w") as f:
-            for res in result:
-                run_id, train_loss, val_loss, test_loss = res
-                if val_loss < best_val_loss:
-                    best_run = run_id
-                    best_train_loss = train_loss
-                    best_val_loss = val_loss
-                    best_test_loss = test_loss
-                f.write("{run_id} {train} {val} {test}\n".format(run_id=run_id,
-                                                                 train=train_loss,
-                                                                 val=val_loss,
-                                                                 test=test_loss))
-        mlflow.log_artifact(results_file_path, "training_runs.txt")
-        # record which run produced the best results, store it as an artifact
-        best_run_path = os.path.join(os.path.join(tmp, "best_run.txt"))
-        with open(best_run_path, "w") as f:
-            f.write("{run_id} {train} {val} {test}\n".format(run_id=best_run,
-                                                             train=best_train_loss,
-                                                             val=best_val_loss,
-                                                             test=best_test_loss))
-        mlflow.log_artifact(best_run_path, "best-run")
-        mlflow.log_metric(val_metric, best_val_loss)
-        mlflow.log_metric(test_metric, best_test_loss)
-        shutil.rmtree(tmp)
+
+        # find the best run, log its metrics as the final metrics of this run.
+        client = MlflowClient()
+        runs = client.search_runs([experiment_id],
+                                  "tags.mlflow.parentRunId = '{run_id}' ".format(
+                                      run_id=run.info.run_id
+                                  ))
+        best_val_train = _inf
+        best_val_valid = _inf
+        best_val_test = _inf
+        best_run = None
+        for r in runs:
+            if r.data.metrics["val_rmse"] < best_val_valid:
+                best_run = r
+                best_val_train = r.data.metrics["train_rmse"]
+                best_val_valid = r.data.metrics["val_rmse"]
+                best_val_test = r.data.metrics["test_rmse"]
+        mlflow.set_tag("best_run", best_run.info.run_id)
+        mlflow.log_metric("train_{}".format(metric), best_val_train)
```

Comment:
We could optionally use `mlflow.log_metrics({"train_{}".format(metric): best_val_train, "val_{}".format(metric): best_val_valid, "test_{}".format(metric): best_val_test})` here as an example of how to log multiple metrics in a single call, but also fine to leave as is.

Reply:
hmm that's a good point. I think we should do it. It's an example, it should show the best of mlflow! ;)